File: src/test/java/com/stratio/connector/hdfs/engine/HDFSMetadataEngineTest.java
package com.stratio.connector.hdfs.engine;

import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.powermock.modules.junit4.PowerMockRunner;

import com.stratio.connector.commons.connection.Connection;
import com.stratio.connector.commons.connection.exceptions.HandlerConnectionException;
import com.stratio.connector.hdfs.connection.HDFSConnectionHandler;
import com.stratio.connector.hdfs.utils.HDFSClient;
import com.stratio.crossdata.common.data.CatalogName;
import com.stratio.crossdata.common.data.ClusterName;
import com.stratio.crossdata.common.data.ColumnName;
import com.stratio.crossdata.common.data.TableName;
import com.stratio.crossdata.common.exceptions.ExecutionException;
import com.stratio.crossdata.common.exceptions.UnsupportedException;
import com.stratio.crossdata.common.metadata.CatalogMetadata;
import com.stratio.crossdata.common.metadata.ColumnMetadata;
import com.stratio.crossdata.common.metadata.TableMetadata;
import com.stratio.crossdata.common.statements.structures.Selector;

@RunWith(PowerMockRunner.class)
public class HDFSMetadataEngineTest {

    private static final ClusterName CLUSTERNAME_CONSTANT =  new ClusterName("cluster_name");
    private static final String CLUSTER_NAME = "cluster_name";
    private static final String CATALOG = "catalog";
    private static final String INDEX_NAME = "INDEX_NAME".toLowerCase();
    private static final String TYPE_NAME = "TYPE_NAME".toLowerCase();
    private TableName tableName = new TableName(INDEX_NAME, TYPE_NAME);

    @Mock
    private HDFSConnectionHandler connectionHandler;
    @Mock
    private Connection<HDFSClient> connection;
    @Mock
    private HDFSClient client;

    @Mock
    private HDFSMetadataEngine hdfsMetadataEngine;

    @Before
    public void before() throws HandlerConnectionException {

        when(connectionHandler.getConnection(CLUSTER_NAME)).thenReturn(connection);
        when(connection.getNativeConnection()).thenReturn(client);
        hdfsMetadataEngine = new HDFSMetadataEngine(connectionHandler);

    }
    @After
    public void after() throws Exception {
    }

    @Test
    public void createCatalogTest() throws UnsupportedException, ExecutionException {

        hdfsMetadataEngine.createCatalog(new ClusterName("cluster_name"),new CatalogMetadata(new CatalogName(CATALOG), null, null));
        verify(client, times(1)).mkdir(anyString());

    }

    @Test
    public void dropCatalogTest() throws UnsupportedException, ExecutionException {

        hdfsMetadataEngine.dropCatalog(new CatalogName(CATALOG), connection);
        verify(client, times(1)).deleteFile(anyString());

    }
    @Test
    public void createTableTest() throws UnsupportedException, ExecutionException {


        TableMetadata tablemetadta = mock(TableMetadata.class);
        Map<Selector, Selector> options = Collections.EMPTY_MAP;
        Map<ColumnName, ColumnMetadata> columns = new HashMap<>();
        Map indexex = Collections.EMPTY_MAP;
        List<ColumnName> partitionKey = Collections.EMPTY_LIST;
        List<ColumnName> clusterKey   = Collections.EMPTY_LIST;

        hdfsMetadataEngine.createTable(new ClusterName("cluster_name"),
                new TableMetadata(tableName, options, columns, indexex, CLUSTERNAME_CONSTANT, partitionKey, clusterKey));

        verify(client, times(1)).addFile(anyString());
    }

    @Test
    public void dropTableTest() throws UnsupportedException, ExecutionException {

        hdfsMetadataEngine.dropTable(tableName,connection);

        verify(client, times(1)).deleteFile(anyString());
    }
}

File: connector-hdfs/src/test/scala/com/stratio/connector/hdfs/engine/HDFSMetadataEngineTest.scala
/*
 *
 *  Licensed to STRATIO (C) under one or more contributor license agreements.
 *  See the NOTICE file distributed with this work for additional information
 *  regarding copyright ownership. The STRATIO (C) licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License. You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied. See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package com.stratio.connector.hdfs.engine

import java.util

import com.stratio.crossdata.common.data.{IndexName, TableName}
import com.stratio.crossdata.common.metadata.{IndexMetadata, ColumnMetadata, TableMetadata, CatalogMetadata}
import com.stratio.crossdata.common.statements.structures.Selector
import org.apache.hadoop.fs.Path
import com.stratio.crossdata.common.data.ColumnName
import org.scalamock.scalatest.MockFactory
import org.scalatest.{Matchers, FlatSpec}

class HDFSMetadataEngineTest extends FlatSpec with Matchers with MockFactory with FSConstants{

  trait MetadataEngineData {

    val metadataEng = new HDFSMetadataEngine(connectionHandler)

    val options = Map[Selector, Selector]()

    val tables = Map[TableName, TableMetadata]()

    val indexes = Map[IndexName,IndexMetadata]()

    val columns = new util.LinkedHashMap[ColumnName, ColumnMetadata]()

    val partitionKey = List[ColumnName]()

    val clusterKey = List[ColumnName]()

    val pathCatalog = new Path(s"$catalogName")

    val pathTable = new Path(s"$catalogName/tablename")

  }

  import scala.collection.JavaConversions._

  behavior of "a Metadata Engine"

  it should "create a new catalog" in new MetadataEngineData {

    metadataEng.createCatalog(new CatalogMetadata(catalogName, options, tables),
      hdfsConnection)

    fakeFileSystem.exists(pathCatalog) should equal (true)

    fakeFileSystem.delete(pathCatalog, true)

    fakeFileSystem.close()

  }

  it should "create a new table" in new MetadataEngineData {

    metadataEng.createTable(new TableMetadata(tableName, options, columns, indexes, clusterName, partitionKey, clusterKey), hdfsConnection)

    fakeFileSystem.exists(pathTable) should equal (true)

    fakeFileSystem.delete(pathCatalog, true)

    fakeFileSystem.close()

  }
}
